ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp

```diff
@@ -206,6 +206,32 @@ Tensor Softplus::invoke(
         optional_output_tensor);
 }
 
+// tanh[x] = (exp[x] - exp[-x]) / (exp[x] - exp[-x])
+Tensor Tanh::invoke(
+    QueueId queue_id,
+    const Tensor& input_tensor,
+    const std::optional<MemoryConfig>& memory_config,
+    const std::optional<Tensor>& optional_output_tensor,
+    bool accuracy) {
+    UnaryOpType op_type = UnaryOpType::TANH;
+    if (!accuracy) {
+        return detail::unary_impl(
+            queue_id, input_tensor, {UnaryWithParam{op_type}}, memory_config, optional_output_tensor);
+    } else {
+        TT_FATAL(
+            input_tensor.get_dtype() == DataType::BFLOAT16,
+            "Supported dtypes for tanh with accuracy mode enabled is : BFLOAT16");
+        Tensor tanh_res = detail::unary_impl(
+            queue_id, input_tensor, {UnaryWithParam{op_type}}, memory_config, optional_output_tensor);
+        Tensor e_pos_x = ttnn::exp(input_tensor, false, memory_config);
+        Tensor e_neg_x = ttnn::exp(ttnn::neg(input_tensor, memory_config), false, memory_config);
+        Tensor nr_term_add = ttnn::add(e_pos_x, e_neg_x, std::nullopt, memory_config);
+        Tensor nr_term_sub = ttnn::subtract(e_pos_x, e_neg_x, std::nullopt, memory_config);
+        Tensor abs_val = ttnn::abs(input_tensor);
+        return ttnn::where(ttnn::gt(abs_val, 3.5f), tanh_res, ttnn::divide(nr_term_sub, nr_term_add));
+    }
```

Comment:
https://github.com/tenstorrent/tt-metal/blob/main/contributing/BestPractices.md#16-use-early-exit-for-contract-checks

```suggestion
    UnaryWithParam op_type = UnaryOpType::TANH;
    if (!accuracy) {
        return detail::unary_impl(
            queue_id, input_tensor, {op_type}, memory_config, optional_output_tensor);
    }

    TT_FATAL(
        input_tensor.get_dtype() == DataType::BFLOAT16,
        "Supported dtypes for tanh with accuracy mode enabled is : BFLOAT16");
    const auto tanh_res = detail::unary_impl(
        queue_id, input_tensor, {op_type}, memory_config, optional_output_tensor);
    const auto e_pos_x = ttnn::exp(input_tensor, false, memory_config);
    const auto e_neg_x = ttnn::exp(ttnn::neg(input_tensor, memory_config), false, memory_config);
    const auto nr_term_add = ttnn::add(e_pos_x, e_neg_x, std::nullopt, memory_config);
    const auto nr_term_sub = ttnn::subtract(e_pos_x, e_neg_x, std::nullopt, memory_config);
    const auto abs_val = ttnn::abs(input_tensor);
    return ttnn::where(ttnn::gt(abs_val, 3.5f), tanh_res, ttnn::divide(nr_term_sub, nr_term_add));
```

Reply:
Also, have you tried using in-place operations?

Reply:
- Since I'm using `ttnn::exp`, it is not accurate for the bigger range. Initially tested that with Andrija's model and PCC went to negative and just unpredictable. Beyond 3.5 , the output for `tanh` is always 1. But using `exp` doesn't give that. The lookup table result is more accurate for the range outside of 3.5. in general, `ttnn::exp` has issues with larger input values.
- Performance will get better when this is implemented as a single device operation rather than composite which is the next step as mentioned in the desc. In a device operation all these intermediate tensors will be intermediate CBs
- I will add the changes for early-return and auto deduction
- I don't see much scope for using in-place operation as it alters the input tensor in memory.. but I want to reuse the tensors like `e_pos_x` , `e_neg_x` .. I can deallocate them after I'm done with calculating the numerator and denominator. But I remember someone saying a deallocate is not necessary.. is it true? I could re-use the input_tensor for taking the absolute value but  I don't want to overwrite the input_tensor since I'm not sure whether it'll be needed later in a model.
- I will try the alternate approach to see if it gives the same PCC.

Reply:
I have tried to reuse the tensors with the simplified approach. It gives similar PCC and Andrija tested it with albert_v2 models as well. It looks good.